sdk/core/azure-core/azure/core/messaging.py

```diff
@@ -0,0 +1,175 @@
+     unique for each distinct event. If not provided, a random UUID will be generated and used.
+    :vartype id: str
+    :ivar extensions: A CloudEvent MAY include any number of additional context attributes
+     with distinct names represented as key - value pairs. Each extension must be alphanumeric, lower cased
+     and must not exceed the length of 20 characters.
+    :vartype extensions: Dict
+    """
+
+    def __init__(self, source, type, **kwargs):  # pylint: disable=redefined-builtin
+        # type: (str, str, **Any) -> None
+        self.source = source  # type: str
+        self.type = type  # type: str
+        self.specversion = kwargs.pop("specversion", "1.0")  # type: str
+        self.id = kwargs.pop("id", str(uuid.uuid4()))  # type: str
+        self.time = kwargs.pop("time", datetime.now(TZ_UTC))  # type: datetime
+
+        _optional_attributes = ["datacontenttype", "dataschema", "subject", "data"]
+
+        for attr in _optional_attributes:
+            val = None if attr not in kwargs else kwargs.pop(attr)
+            setattr(self, attr, val)
+
+        try:
+            self.extensions = kwargs.pop("extensions") # type: Dict
+            for key in self.extensions.keys():
+                if not key.islower() or not key.isalnum():
+                    raise ValueError(
+                        "Extension attributes should be lower cased and alphanumeric."
+                    )
+        except KeyError:
+            self.extensions = cast(Dict, None)
+
+        if kwargs:
+            remaining = ", ".join(kwargs.keys())
+            raise ValueError(
+                "Unexpected keyword arguments {}. Any extension attributes must be passed explicitly using extensions."
+                .format(remaining)
+            )
+
+    def __repr__(self):
+        return "CloudEvent(source={}, type={}, specversion={}, id={}, time={})".format(
+            self.source, self.type, self.specversion, self.id, self.time
+        )[:1024]
+
+    @classmethod
+    def from_dict(cls, event):
+        # type: (Dict) -> CloudEvent
+        """
+        Returns the deserialized CloudEvent object when a dict is provided.
+        :param event: The dict representation of the event which needs to be deserialized.
+        :type event: dict
+        :rtype: CloudEvent
+        """
+        kwargs = {} # type: Dict[Any, Any]
+        reserved_attr = [
+            "data",
+            "data_base64",
+            "id",
+            "source",
+            "type",
+            "specversion",
+            "time",
+            "dataschema",
+            "datacontenttype",
+            "subject",
+        ]
+
+        if "data" in event and "data_base64" in event:
+            raise ValueError(
+                "Invalid input. Only one of data and data_base64 must be present."
+            )
+        if 'data' not in event and 'data_base64' not in event:
+            kwargs.setdefault("data", None)
```

Comment:
Why are you using `setdefault` here? It is generally used when you get a partially filled out `dict` and you don't want to overwrite existing values. But you are defining `kwargs` in the method (i.e. it is not a parameter passed in to the method).

Please rename, and use direct assignment (e.g. `eventargs['data'] = None)

Reply:
...also, do you really need to set `"data"` to `None` here? The `dict` passed in didn't have it, so you should be able to create the CloudEvent instance without passing in the keyword argument, right?